File: lib/utils/input_sanitizer.dart
/// Input sanitization utilities to prevent XSS, SQL injection, and other attacks
class InputSanitizer {
  static final _htmlTagPattern = RegExp(r'<[^>]*>');
  static final _scriptPattern = RegExp(r'<script[^>]*>.*?</script>', caseSensitive: false);
  static final _sqlInjectionPatterns = [
    RegExp(r';\\s*--'),
    RegExp(r'union\\s+select', caseSensitive: false),
    RegExp(r'drop\\s+table', caseSensitive: false),
    RegExp(r'insert\\s+into', caseSensitive: false),
    RegExp(r'delete\\s+from', caseSensitive: false),
  ];
  static final _xssPatterns = [
    RegExp(r'javascript:', caseSensitive: false),
    RegExp(r'on\\w+\\s*=', caseSensitive: false),
    RegExp(r'<iframe', caseSensitive: false),
    RegExp(r'<embed', caseSensitive: false),
    RegExp(r'<object', caseSensitive: false),
  ];

  /// Sanitize text input by removing dangerous patterns
  static String sanitizeText(String? input) {
    if (input == null || input.isEmpty) return '';
    
    String sanitized = input.trim();
    
    // Remove script tags
    sanitized = sanitized.replaceAll(_scriptPattern, '');
    
    // Remove HTML tags
    sanitized = sanitized.replaceAll(_htmlTagPattern, '');
    
    // Remove null bytes
    sanitized = sanitized.replaceAll('\x00', '');
    
    return sanitized;
  }

  /// Check if input contains malicious patterns
  static bool containsMaliciousPattern(String? input) {
    if (input == null || input.isEmpty) return false;
    
    final lower = input.toLowerCase();
    
    // Check SQL injection patterns
    for (final pattern in _sqlInjectionPatterns) {
      if (pattern.hasMatch(lower)) return true;
    }
    
    // Check XSS patterns
    for (final pattern in _xssPatterns) {
      if (pattern.hasMatch(input)) return true;
    }
    
    return false;
  }

  /// Sanitize and validate email
  static String? sanitizeEmail(String? email) {
    if (email == null || email.isEmpty) return null;
    
    final sanitized = email.trim().toLowerCase();
    
    if (containsMaliciousPattern(sanitized)) return null;
    if (sanitized.length > 254) return null;
    
    return sanitized;
  }

  /// Sanitize name (allow letters, spaces, hyphens, apostrophes)
  static String? sanitizeName(String? name) {
    if (name == null || name.isEmpty) return null;
    
    String sanitized = name.trim();
    
    if (containsMaliciousPattern(sanitized)) return null;
    if (sanitized.length > 100) return null;
    
    // Remove any characters that aren't letters, spaces, hyphens, or apostrophes
    sanitized = sanitized.replaceAll(RegExp(r"[^a-zA-Z\s\-']"), '');
    
    return sanitized.isEmpty ? null : sanitized;
  }

  /// Sanitize phone number (keep only digits, +, -, (, ), spaces)
  static String? sanitizePhone(String? phone) {
    if (phone == null || phone.isEmpty) return null;
    
    final sanitized = phone.trim();
    
    if (containsMaliciousPattern(sanitized)) return null;
    
    // Keep only valid phone characters
    final cleaned = sanitized.replaceAll(RegExp(r'[^\d\s\+\-\(\)]'), '');
    
    return cleaned.isEmpty ? null : cleaned;
  }

  /// Sanitize general text fields (bio, description, etc.)
  static String? sanitizeTextField(String? text, {int maxLength = 500}) {
    if (text == null || text.isEmpty) return null;
    
    String sanitized = sanitizeText(text);
    
    if (containsMaliciousPattern(sanitized)) return null;
    if (sanitized.length > maxLength) {
      sanitized = sanitized.substring(0, maxLength);
    }
    
    return sanitized.isEmpty ? null : sanitized;
  }

  /// Sanitize URL
  static String? sanitizeUrl(String? url) {
    if (url == null || url.isEmpty) return null;
    
    final sanitized = url.trim();
    
    if (containsMaliciousPattern(sanitized)) return null;
    if (sanitized.length > 2000) return null;
    if (!sanitized.startsWith('http://') && !sanitized.startsWith('https://')) {
      return null;
    }
    
    return sanitized;
  }

  /// Sanitize search query
  static String? sanitizeSearchQuery(String? query) {
    if (query == null || query.isEmpty) return null;
    
    String sanitized = query.trim();
    
    if (containsMaliciousPattern(sanitized)) return null;
    if (sanitized.length > 100) {
      sanitized = sanitized.substring(0, 100);
    }
    
    // Allow only safe characters for search
    sanitized = sanitized.replaceAll(RegExp(r"[^a-zA-Z0-9\s\-\.,']"), '');
    
    return sanitized.isEmpty ? null : sanitized;
  }

  /// Sanitize integer input
  static int? sanitizeInt(dynamic value, {int? min, int? max}) {
    if (value == null) return null;
    
    final parsed = int.tryParse(value.toString());
    if (parsed == null) return null;
    
    if (min != null && parsed < min) return null;
    if (max != null && parsed > max) return null;
    
    return parsed;
  }

  /// Sanitize map of user input (for API requests)
  static Map<String, dynamic> sanitizeMap(Map<String, dynamic> input) {
    final sanitized = <String, dynamic>{};
    
    for (final entry in input.entries) {
      final key = entry.key;
      final value = entry.value;
      
      if (value == null) continue;
      
      if (value is String) {
        final cleaned = sanitizeText(value);
        if (cleaned.isNotEmpty && !containsMaliciousPattern(cleaned)) {
          sanitized[key] = cleaned;
        }
      } else if (value is int || value is double || value is bool) {
        sanitized[key] = value;
      } else if (value is List) {
        sanitized[key] = value.map((e) => 
          e is String ? sanitizeText(e) : e
        ).toList();
      } else if (value is Map) {
        sanitized[key] = sanitizeMap(value as Map<String, dynamic>);
      }
    }
    
    return sanitized;
  }
}

File: supabase/functions/_shared/input_sanitizer.ts
// Input Sanitization and Validation Utilities

export interface SanitizationResult {
  isValid: boolean;
  sanitized?: string;
  error?: string;
}

/**
 * Sanitize string for logging to prevent log injection
 */
export function sanitizeForLog(input: string, maxLength: number = 100): string {
  if (!input) return '';
  return input
    .replace(/[\n\r\t]/g, ' ')
    .replace(/[^\x20-\x7E]/g, '')
    .substring(0, maxLength);
}

/**
 * Validate and sanitize email
 */
export function sanitizeEmail(email: string): SanitizationResult {
  if (!email || typeof email !== 'string') {
    return { isValid: false, error: 'Email is required' };
  }

  const trimmed = email.trim().toLowerCase();

  if (trimmed.length > 254) {
    return { isValid: false, error: 'Email too long' };
  }

  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  if (!emailRegex.test(trimmed)) {
    return { isValid: false, error: 'Invalid email format' };
  }

  return { isValid: true, sanitized: trimmed };
}

/**
 * Validate and sanitize phone number
 */
export function sanitizePhone(phone: string): SanitizationResult {
  if (!phone || typeof phone !== 'string') {
    return { isValid: false, error: 'Phone number is required' };
  }

  const cleaned = phone.replace(/\D/g, '');

  if (cleaned.length < 10 || cleaned.length > 15) {
    return { isValid: false, error: 'Invalid phone number length' };
  }

  return { isValid: true, sanitized: cleaned };
}

/**
 * Sanitize text input (names, titles, etc.)
 */
export function sanitizeText(
  text: string,
  minLength: number = 2,
  maxLength: number = 100
): SanitizationResult {
  if (!text || typeof text !== 'string') {
    return { isValid: false, error: 'Text is required' };
  }

  const trimmed = text.trim();

  if (trimmed.length < minLength) {
    return { isValid: false, error: `Text too short (min: ${minLength})` };
  }

  if (trimmed.length > maxLength) {
    return { isValid: false, error: `Text too long (max: ${maxLength})` };
  }

  // Remove potentially dangerous characters
  const sanitized = trimmed.replace(/[<>\"']/g, '');

  return { isValid: true, sanitized };
}

/**
 * Validate UUID format
 */
export function isValidUUID(uuid: string): boolean {
  if (!uuid || typeof uuid !== 'string') return false;
  
  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  return uuidRegex.test(uuid);
}

/**
 * Sanitize search query
 */
export function sanitizeSearchQuery(query: string): SanitizationResult {
  if (!query || typeof query !== 'string') {
    return { isValid: false, error: 'Search query is required' };
  }

  const trimmed = query.trim();

  if (trimmed.length < 2) {
    return { isValid: false, error: 'Search query too short' };
  }

  if (trimmed.length > 100) {
    return { isValid: false, error: 'Search query too long' };
  }

  // Remove SQL injection attempts and special characters
  const sanitized = trimmed
    .replace(/[;'"\\]/g, '')
    .replace(/--/g, '')
    .replace(/\/\*/g, '')
    .replace(/\*\//g, '');

  return { isValid: true, sanitized };
}

/**
 * Validate integer within range
 */
export function validateInteger(
  value: any,
  min: number,
  max: number
): SanitizationResult {
  const num = parseInt(value);

  if (isNaN(num)) {
    return { isValid: false, error: 'Invalid number' };
  }

  if (num < min || num > max) {
    return { isValid: false, error: `Number must be between ${min} and ${max}` };
  }

  return { isValid: true, sanitized: num.toString() };
}

/**
 * Validate date string
 */
export function validateDate(dateStr: string): SanitizationResult {
  if (!dateStr || typeof dateStr !== 'string') {
    return { isValid: false, error: 'Date is required' };
  }

  const date = new Date(dateStr);

  if (isNaN(date.getTime())) {
    return { isValid: false, error: 'Invalid date format' };
  }

  return { isValid: true, sanitized: date.toISOString() };
}
